### ISSUE TITLE
Consider Keycloak resources for authorization instead of dynamically created roles

### ISSUE THREAD

**Issue by mmurto** [eclipse-apoapsis/ort-server#984]
As a warning, I'm not super deep on Keycloak. Nevertheless, the way ORT Server uses Keycloak's client roles dynamically (`permission_organization_$ORGID_write` etc.) seems unintuitive. What ORT Server wants to do is to protect resources (organizations, products etc.), and for that Keycloak has [native functionality to manage resources, policies and permissions](https://www.keycloak.org/docs/latest/authorization_services/index.html#_resource_overview). To me this seems like a better fit than the current method of handling authorization. There have also been some reports where [large number of roles has led to performance issues](https://github.com/keycloak/keycloak/issues/12885), and with the current authorization model the role count grows very large.

What is the reason for going with these dynamic roles for authorization instead of Keycloak's resource authorization service?

**Comment by mnonnenmacher** [comment]
I totally agree with that. I wanted to look into resources since the initial implementation but never found the time.

@mmurto In the context of #504, do you know if the other auth providers you suggest there have similar features?

**Comment by sschuberth** [comment]
> I totally agree with that. I wanted to look into resources since the initial implementation but never found the time.

Is this by now on your agenda @mnonnenmacher?

**Comment by mnonnenmacher** [comment]
> > I totally agree with that. I wanted to look into resources since the initial implementation but never found the time.
> 
> Is this by now on your agenda @mnonnenmacher?

It's low priority to me because the current implementation works and there are other more important tasks.

**Comment by mnonnenmacher** [comment]
I am considering if it would be a better solution to move the role management into the database. This would make it permission handling a lot easier, for example, when doing a paginated query for organizations, they could already be filtered by the database query. This would also simplify the interface to the auth provider and make #504 easier.

@mmurto Do you have an opinion on that?

**Comment by mmurto** [comment]
I think it would likely be a good solution. To me it seems that if more fine-graned authorization is needed than group roles, the ecosystem seems to gravitate towards using a separate service for that, like [OpenFGA](https://github.com/openfga), [Ory Keto](https://github.com/ory/keto) and others. While those could provide some benefits, I'm not sure if our use-case and scale warrant a separate service for just authorization.

What is a part of most of the authentication providers, like Keycloak v25 onwards, is organization support. The providers often provide some general functionality, like member management, invitation emails etc. for organizations, so having that managed by the authentication service could be worthwhile, though I'm not sure how it should be architected.

**Comment by mmurto** [comment]
> > What is a part of most of the authentication providers, like Keycloak v25 onwards, is organization support. The providers often provide some general functionality, like member management, invitation emails etc. for organizations, so having that managed by the authentication service could be worthwhile, though I'm not sure how it should be architected.
> 
> If I understand this correctly this is mostly about user management and high level grouping of users?

Yeah, things like sending invitation emails to new users in an organization, connecting external auth providers for single sign on to automatically add new users logging in with the company's Google Workpace etc. to their ORT Server organization, and enabling organization administrators to force all of the organization's users to use MFA, etc.

One way to sync the organization of the provider with organization permissions stored in ORT Server's database would be to add a webhook endpoint, as a lot of the providers allow for webhooks for events like new user etc.
